charm/main.py

```diff
@@ -0,0 +1,218 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charm
+
+
+import argparse
+import logging
+import os
+import sys
+
+from charm import logsetup
+
+logger = logging.getLogger(__name__)
+
+
+class CommandError(Exception):
+    """Base exception for all error commands.
+
+    It optionally receives an `retcode` parameter that will be the returned code
+    by the process on exit.
+    """
+    def __init__(self, message, retcode=-1):
+        self.retcode = retcode
+        super().__init__(message)
+
+
+class BaseCommand:
+
+    name = None
+    help_msg = None
+
+    def __init__(self, group):
+        if self.name is None or self.help_msg is None:
+            raise RuntimeError("Command not properly created: {!r}".format(self.__class__))
+        self.group = group
+
+    def fill_parser(self, parser):
+        """Overwrite in each command and fill the parser with command-specific parameters.
+
+        If not overwritten, the command will not have any parameters.
+        """
+
+
+class VersionCommand(BaseCommand):
+    """Show the version."""
+    name = 'version'
+    help_msg = "show the version"
+
+    def run(self, parsed_args):
+        """Run the command."""
+        # XXX: we need to define how we want to store the project version (in config/file/etc.)
+        version = '0.0.1'
+        logger.info('Version: %s', version)
+
+
+# XXX: while VersionCommand above probably would stay here, the other commands surely will
+# be located in different files... that said, I'm keeping these examples here, for
+# simplicitiy
+
+
+class CommandExampleDebug(BaseCommand):
+    """Just an example."""
+
+    name = 'example-debug'
+    help_msg = "show msg in debug"
+
+    def fill_parser(self, parser):
+        parser.add_argument('foo')
+        parser.add_argument('--bar', action='store_true', help="To use this command in a bar")
+
+    def run(self, parsed_args):
+        logger.debug(
+            "Example showing log in DEBUG: foo=%s bar=%s", parsed_args.foo, parsed_args.bar)
+
+
+class CommandExampleError(BaseCommand):
+    """Just an example."""
+
+    name = 'example-error'
+    help_msg = "show msg in error"
+
+    def run(self, parsed_args):
+        logger.error("Example showing log in ERROR.")
+
+
+# Collect commands in different groups, for easier human consumption. Note that this is not
+# declared in each command because it's much easier to do this separation/grouping in one
+# central place and not distributed in several classes/files.
+COMMANDS_GROUPS = [
```

Comment:
is this exported on purpose?
(same question for `CustomArgumentParser` and `Dispatch`; I understand that the commands themselves would be exported because we'll be importing them from main when they get moved to their own place and renaming them when they get moved is strange)

Reply:
This is a CLI tool, not at all thought as "a module": we're not exporting functionality nor we have an API for consumption as a module, so it doesn't really apply the public/private reasoning...